libavcodec/vaapi_encode.c

```diff
@@ -1845,6 +1845,42 @@ static av_cold int vaapi_encode_init_gop_structure(AVCodecContext *avctx)
         ref_l1 = attr.value >> 16 & 0xffff;
     }
 
+#if VA_CHECK_VERSION(1, 9, 0)
+    attr = (VAConfigAttrib) { VAConfigAttribPredictionDirection };
+    vas = vaGetConfigAttributes(ctx->hwctx->display,
+                                ctx->va_profile,
+                                ctx->va_entrypoint,
+                                &attr, 1);
+    if (vas != VA_STATUS_SUCCESS) {
+        av_log(avctx, AV_LOG_WARNING, "Failed to query prediction direction "
+               "attribute: %d (%s).\n", vas, vaErrorStr(vas));
+        if (ctx->p_to_gpb) {
+            ctx->p_to_gpb = 0;
+            av_log(avctx, AV_LOG_WARNING, "Invalid p_to_gpb setting, turn to "
+                   "use regular P frames.\n");
+        }
+    } else if (attr.value == VA_ATTRIB_NOT_SUPPORTED) {
+        if (ctx->p_to_gpb) {
+            ctx->p_to_gpb = 0;
+            av_log(avctx, AV_LOG_WARNING, "Invalid p_to_gpb setting, turn to "
+                   "use regular P frames.\n");
+        }
+    } else {
+        if (attr.value & VA_PREDICTION_DIRECTION_BI_NOT_EMPTY) {
+            if (!ctx->p_to_gpb) {
+                ctx->p_to_gpb = 1;
+                av_log(avctx, AV_LOG_WARNING, "Invalid p_to_gpb setting, turn to "
+                       "use GPB B frames.\n");            }
+        } else {
+            if (ctx->p_to_gpb) {
+                ctx->p_to_gpb = 0;
+                av_log(avctx, AV_LOG_WARNING, "Invalid p_to_gpb setting, turn to "
+                    "use regular P frames.\n");
```

Comment:
According the logic here, it seems we needn't provide the option and FFmpeg should respect driver's capability to set P or GPB frame.

Reply:
OK, I will remove the option, just make it always as auto correct mode.